File: web/spring-webflux/webflux-fn/src/main/java/com/example/UserHandler.java
package com.example;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.web.reactive.function.BodyInserters.fromValue;
import static org.springframework.web.reactive.function.server.RequestPredicates.accept;
import static org.springframework.web.reactive.function.server.RequestPredicates.contentType;
import static org.springframework.web.reactive.function.server.ServerResponse.badRequest;
import static org.springframework.web.reactive.function.server.ServerResponse.created;
import static org.springframework.web.reactive.function.server.ServerResponse.notFound;
import static org.springframework.web.reactive.function.server.ServerResponse.ok;

import java.net.URI;

import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.RouterFunctions;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;

import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Component
@RequiredArgsConstructor
public class UserHandler {
    private final UserRepository userRepository;

    @Bean
    public RouterFunction<ServerResponse> routes() {
        return RouterFunctions.route()
                              .path("/users", builder -> builder
                                      .GET("", accept(APPLICATION_JSON), this::findAll)
                                      .GET("/{id}", accept(APPLICATION_JSON), this::findById)
                                      .POST("", contentType(APPLICATION_JSON), this::create))
                              .build();
    }

    private Mono<ServerResponse> findAll(ServerRequest request) {
        Flux<User> users = request.queryParam("name")
                                  .map(userRepository::findByName)
                                  .orElseGet(userRepository::findAll);
        return ok().body(users, User.class);
    }

    private Mono<ServerResponse> findById(ServerRequest request) {
        String id = request.pathVariable("id");
        return userRepository.findById(id)
                             .flatMap(user -> ok().body(fromValue(user)))
                             .switchIfEmpty(notFound().build());
    }

    private Mono<ServerResponse> create(ServerRequest request) {
        Mono<User> user = request.bodyToMono(User.class);
        return userRepository.saveAll(user)
                             .single()
                             .flatMap(created -> created(URI.create("/users/" + created.getId())).build())
                             .switchIfEmpty(badRequest().build());
    }
}

File: kotlin/coroutines/webflux-fn/src/main/kotlin/com/example/UserHandler.kt
package com.example

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.ServerResponse.ok
import org.springframework.web.reactive.function.server.bodyAndAwait
import org.springframework.web.reactive.function.server.bodyValueAndAwait
import org.springframework.web.reactive.function.server.coRouter

@Component
class UserHandler {
    @Bean
    fun routes() = coRouter {
        "/users".nest {
            GET("", ::findAll)
            GET("/{id}", ::findById)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    suspend fun findAll(request: ServerRequest): ServerResponse {
        val users: Flow<User> = (1..5).map { User(it, "name$it") }.asFlow()
        return ok().bodyAndAwait(users)
    }

    suspend fun findById(request: ServerRequest): ServerResponse {
        delay(1000)
        val id = request.pathVariable("id").toInt()
        return ok().bodyValueAndAwait(User(id, "name$id"))
    }
}
